### ISSUE TITLE
DSX uses a lot more memory, slower to launch

### ISSUE THREAD

**Issue by decentjohn** [decentespresso/de1app#163]
I am not certain that all these repeated-image-loading in DSX is the cause of the memory bloat, but DSX certainly does use a lot more memory than Insight or Metric (about 70% more - ie 120mb more)

On OSX, just loading the app:

Insight:
![insight](https://user-images.githubusercontent.com/11965250/120097076-2fa0fc00-c161-11eb-92d8-b9d82e0159c2.jpg)

Metric:
![metric](https://user-images.githubusercontent.com/11965250/120097078-30d22900-c161-11eb-807c-e9e033fd9fab.jpg)

DSX:
![dsx](https://user-images.githubusercontent.com/11965250/120097079-316abf80-c161-11eb-841c-c5808765a3dc.jpg)

My guess is that all these loads of the same image is what's causing the problem:

DSx_skin.tcl:add_de1_image "DSx_3_coffee" 400 410 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 880 410 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 400 620 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 880 620 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 400 830 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 880 830 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 1880 410 "[skin_directory_graphics]/icons/click1.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 530 1210 "[skin_directory_graphics]/icons/button4.png"
DSx_skin.tcl:add_de1_image "DSx_3_coffee" 780 1210 "[skin_directory_graphics]/icons/button4.png"

I looked into it, and this add_de1_image is now a dui function, so I prefer to leave it to @ebengoechea to fix this, and not mess with his code.

There are two solutions I'd recommend, and de1app currently implements both for Insight skin's loading of images:

- lazy loading -- images are not loaded until they are actually needed (ie, about to be displayed)
- references -- if the same PNG/JPG is used, the reference to the first loaded variable is used, instead of reloading the image into memory

If you look into add_de1_page{}

You'll see
set ::delayed_image_load($name) $pngfilename

to implement the lazy loading.

**Comment by ebengoechea** [comment]
@decentjohn @damian-au
I have DUI now remembering which image files have already been loaded when calling 'dui add image' or 'add_de1_image', and reusing the loaded image if it has been. But this caching does **not** sensibly reduce the memory usage of DSx vs other skins. 

My guess is that this happens because 'add_de1_image' is being used in DSx for small images (buttons and the like), so the image caching helps but is not big in size, and most memory consumption is instead due to the repeated loading of the same page background image (a black rectangle), once per page. I think this is done in DSx proc DSx_plug_bg? and it circunvents both the standard page image loading (that caches background images) as well as the new image caching. So @damian-au would need to change how background page images are loaded (or, as I suggested some time ago, remove background images alltogether, as they are single-color and only add memory overhead).

I'll investigate image delayed load for faster startup tomorrow.

**Comment by ebengoechea** [comment]
@decentjohn @Damian-AU
With my next PR, DUI implements both image caching and image preloading for all images (previously it worked only for page backgrounds, and still I haven't touched that code). Delaying image loading is controlled through the new "preload_images" dui setting, which is initialized to the value of the app setting "preload_all_page_images". 

Nevertheless, on my testing on PC this does not sensibly improve neither loading times nor memory usage of DSx. I got around 7 seconds startup time irrespective of whether the setting is 0 or 1 (and I have double-checked that the image preloading/delaying is working), vs around 3-4 seconds for Insight. FWIW Insight loading times don't change much either when preloading vs delaying (like 3 vs 4 seconds).

So, unless I've made some coding mistake I could not spot (and I've done a lot of tests to be sure it was working as intended), the culprit of loading times & memory usage of DSx is not on the images (at least not in a way that can be solved from DUI).

**Comment by decentjohn** [comment]
There likely is still delayed loading going on with Insight.  page_show refers to it.

w/o delayed loading, Insight used to take about 15s to load.

**Comment by decentjohn** [comment]
Also, memory usage of the app with Insight should be dramatically different if delayed loading of all pages is really disabled.

**Comment by ebengoechea** [comment]
@decentjohn 
Insight is not affected whatsoever by the changes as it only uses images for background and it still uses your code to handle that, not DUI code (when there's a background image, ::add_de1_page is still called, not yet migrated to DUI. This is only used by Insight). **On my PC with an SSD** the loading times of Insight don't change much when you change the value of global variable "preload_all_page_images", and that should have nothing to do with my changes as that is still your code. Maybe it made a big difference on android 5 tablets, but I just don't see it on PC.

What I've done is apply a similar logic (caching & delaying) to all images, not only page background ones. DSx uses the same all-black background image for all of its pages (and I think it is already loaded once, if I read correctly @Damian-AU proc **load_theme**), plus several small images for buttons and icons, which is what the changes I've done apply to.

> Also, memory usage of the app with Insight should be dramatically different if delayed loading of all pages is really disabled.

I haven't said otherwise.

**Comment by decentjohn** [comment]
It's possible/likely that "preload_all_page_images" is an optimization now, that has been superseded by another optimisation.

It's strange that DSX is using so much memory, and I'd like to think it's the images (it was for insight).

Out of curiosity, if you #comment out the image loading in DSX, so that it runs image-free, does the memory footprint shrink hugely? That'd be the test I'd try. But this might be for Damian to do, not you.

**Comment by Damian-AU** [comment]
I played with this a few days ago, not loading the background images does reduce RAM usage by ~100MB, bring it's RAM usage lower than the bar bones Insight instal I had running along side to compare.
I can draw colour shapes for backgrounds, but it will require me to pretty well re-wright most of my pages, when I tried it before, shapes were being created slower than what images where loading, which mean often the shaped would cover other content. 
So I reverted back to images. I don't know how much reduction in RAM shapes will gain, if any. 

DSx loads more than double the amount of pages any other skin does, so I don't think it is unreasonable to expect it to use more RAM.
There are procedures or functionality I can remove from DSx, mainly to do with styling options, that may reduce the load some, but I guess my question is...
Is there an issue?

**Comment by ebengoechea** [comment]
@Damian-AU That sounds strange to me. You're actually using the same single-color background image for all of your pages, so in theory you should only need to load it once (and reuse it) and then it should use much less than 100 MB.

Right now all pages in MimojaCafe, Metric, and DYE use rectangles for background and buttons composed of canvas shapes, without any noticeable penalty in memory or speed.

**Comment by ebengoechea** [comment]
As Damian says, DSx has more content than other skins. More pages per se shouldn't add much overhead, but it uses many more variables and, especially, many more BLT objects than other skins.

**Comment by ebengoechea** [comment]
By the way, on my tests the new caching of images in DUI **did reduce** DSx memory consumption, from around 250 MB to 220 MB on my PC. Only it didn't brought DSx to the memory usage of other skins, as it only applied to the smaller icons & buttons.

**Comment by decentjohn** [comment]
@Damian-AU is it possible for you to do what Insight skin does, namely to delay loading of a background image, until the user actually goes to that page?

The advantages of this are:
- much faster app start
- lower memory profile for most users, because they won't go to all pages.

Disadvantage:
- slightly slower initial page load, but I've not been able to notice it

My current page_show code automatically supports delayed loading. If you use this (I am guessing you do) then filling ::delayed_image_load() up with references to filenames, might be enough to make it work.  If you use something else, then grep for delayed_image_load to see how I do it, and copy/paste the same code.

Here is the code in gui.tcl that does this.  

```
	foreach name $names {
		.can create image {0 0} -anchor nw  -tag [list pages $name] -state hidden 
		if {$::settings(preload_all_page_images) == 1} {
			#.can itemconfigure $names -image $names 
			.can itemconfigure $name -image $names
		} else {
			set ::delayed_image_load($name) $pngfilename
		}
	}
```

**Comment by Damian-AU** [comment]
DSx is now using < 130MB fully load at start up. 
I will update to GitHub with new version later, I want to try improve history list boxes/scrollbars first

**Comment by decentjohn** [comment]
The latest DSX version I'm using on my mac, is now 100mb slimmer when running than when I reported it, and only 15mb larger than Insight.  That's great, and so I'm closing this ticket.

![dsx](https://user-images.githubusercontent.com/11965250/123574873-39229000-d803-11eb-84ae-90e5a88598c7.jpg)
![insight](https://user-images.githubusercontent.com/11965250/123574879-3a53bd00-d803-11eb-9f53-79942d71bc90.jpg)
